### PR TITLE
Add Planck and Rosseland means

### LINKED ISSUE #21: Add Planck and Rosseland mean opacities
Opacities are currently only available in per-frequency form.

### PR DESCRIPTION
**Pull request by brryan**
Frequency-integrated methods require opacities in some frequency-averaged sense. This PR provides Planck and Rosseland mean opacities through the interface. Not all existing opacity models are supported right now, I want to at least get `BRT` though.

### CODE REVIEW

**Comment by Yurlungur** [comment]
It occurs to me that we could compute frequency-averaged opacities numerically if we want to. Should we? At runtime, we probably don't want to... it's probably slow... but we could imagine numerically computing them and then tabulating them in Opac Burrows or something.

**Comment by brryan** [comment]
> It occurs to me that we could compute frequency-averaged opacities numerically if we want to. Should we? At runtime, we probably don't want to... it's probably slow... but we could imagine numerically computing them and then tabulating them in Opac Burrows or something.

Thinking more about this, I think in general we probably do want to do this -- even for "simple" analytic opacities like BRT I don't think the integrals are tractable. I still want to think a bit about where exactly this should be done... there's a "free parameter" in the number of temperature support points used, but maybe letting the downstream application set this isn't worth the cost of a potentially slow numerical integral at initialization. 

EDIT: I think these integrals have to be computed when the opacity object is created because they depend on the arguments passed to the constructor. We could create a separate `mean_opac` class (to avoid computing mean opacities if you don't need them) that takes an `opac` object and some resolution parameters and tabulates the mean opacities.

**Comment by Yurlungur** [comment]
> EDIT: I think these integrals have to be computed when the opacity object is created because they depend on the arguments passed to the constructor. We could create a separate mean_opac class (to avoid computing mean opacities if you don't need them) that takes an opac object and some resolution parameters and tabulates the mean opacities.

I like this idea.

**Comment by brryan** [comment]
@Yurlungur OK I think this is in an OK place now for review, and merge if you're happy with it. Works with non-CGS units, but without the explicit `Variant` class from the regular opacities. No photon support yet.

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+#include <singularity-opac/base/radiation_types.hpp>
+#include <singularity-opac/constants/constants.hpp>
+#include <spiner/databox.hpp>
+
+namespace singularity {
+namespace neutrinos {
+
+using pc = PhysicalConstants<CGS>;
+
+#define EPS (10.0 * std::numeric_limits<Real>::min())
```
Comment (brryan) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp:32]: The old `#define EPS (10.0 * std::numeric_limits<Real>::epsilon())` doesn't work for small numbers, e.g. a typical neutrino opacity of `1.e-20 cm^2/g`. I changed this everywhere.

Comment (Yurlungur) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp:32]: That seems fine. Is there a reason we did `#define` and not `constexpr Real`? The latter seems better...

Comment (Yurlungur) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp:32]: also should we get rid of the 10?

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            lkappaPlanck_(iRho, iT, iYe, idx) = lkappaPlanck;
+            lkappaRosseland_(iRho, iT, iYe, idx) = lkappaRosseland;
+          }
+        }
+      }
+    }
+  }
+
+#ifdef SPINER_USE_HDF
+  SpinerOpacity(const std::string &filename)
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: This should be `MeanOpacity`. Might be worth testing this capability works.

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    hid_t file =
+        H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+    status += lkappaPlanck_.saveHDF(file, SP5::Opac::PlanckMeanOpacity);
+    status += lkappaRosseland_.saveHDF(file, SP5::Opac::RosselandMeanOpacity);
+    status += H5Fclose(file);
+
+    if (status != H5_SUCCESS) {
+      OPAC_ERROR("neutrinos::MeanOpacity: HDF5 error\n");
+    }
+  }
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Nice. Caching mean opacities in file seems useful.

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+#endif
+
+  PORTABLE_INLINE_FUNCTION int nlambda() const { return nlambda_; }
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: None of the mean opacity functions have a `lambda` parameter. I would either add one, or remove this call. I guess it's worth thinking about whether we want a `lambda` input param to these functions. I suspect not, but I don't know.

File: singularity-opac/neutrinos/non_cgs_neutrinos.hpp [hunk 5]
```diff
@@ -234,6 +242,58 @@ class NonCGSUnits {
...
+    // thus multiplication converts (1/cm) to unit system.
+    return alpha * length_unit_;
+  }
+
+ private:
+  MeanOpac mean_opac_;
+  Real time_unit_, mass_unit_, length_unit_, temp_unit_;
+  Real rho_unit_;
+};
+
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/non_cgs_neutrinos.hpp:296]: Is this class needed? Could you put the units directly in the mean opacity class? Do you think we'll be adding more mean opacity types? It doesn't matter much either way, I guess. But the modifier pattern might not be needed for mean opacities. I leave it up to you.

File: singularity-opac/neutrinos/spiner_opac_neutrinos.hpp [hunk 6]
```diff
@@ -61,7 +61,7 @@ enum class DataStatus { Deallocated, OnDevice, OnHost };
 template <typename ThermalDistribution>
 class SpinerOpacity {
  public:
-  static constexpr Real EPS = 10.0 * std::numeric_limits<Real>::epsilon();
+  static constexpr Real EPS = 10.0 * std::numeric_limits<Real>::min();
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/spiner_opac_neutrinos.hpp:64]: should we remove the 10?

File: singularity-opac/neutrinos/thermal_distributions_neutrinos.hpp [hunk 7]
```diff
@@ -27,7 +27,7 @@ namespace neutrinos {
 
 using pc = PhysicalConstants<CGS>;
 
-#define EPS (10.0 * std::numeric_limits<Real>::epsilon())
+#define EPS (10.0 * std::numeric_limits<Real>::min())
```
Comment (Yurlungur) [inline singularity-opac/neutrinos/thermal_distributions_neutrinos.hpp:30]: constexpr?

Comment (Yurlungur) [reply singularity-opac/neutrinos/thermal_distributions_neutrinos.hpp:30]: Maybe `EPS` should be defined in `neutrinos/base` or something, so we don't have the same line floating around everywhere.

File: singularity-opac/photons/gray_opacity_photons.hpp [hunk 8]
```diff
@@ -43,37 +43,37 @@ class GrayOpacity {
   inline void Finalize() noexcept {}
 
   PORTABLE_INLINE_FUNCTION
-  Real AbsorptionCoefficientPerNu(const Real rho, const Real temp,
-                                  const Real nu, Real *lambda = nullptr) const {
+  Real AbsorptionCoefficient(const Real rho, const Real temp, const Real nu,
+                             Real *lambda = nullptr) const {
```
Comment (Yurlungur) [inline singularity-opac/photons/gray_opacity_photons.hpp:47]: Why this change? Do we no longer include the bolumetric/gray absorption coefficient in each opacity class?

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Overall looks good. A few comments below.

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 4 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Well the `lambda` parameter can still be provided to the constructor that calls the non-mean opacities. It's just that lambda is assumed constant for all `rho`, `T`, `Ye`, `nu`.

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 2 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Nice catch! Cmake not currently working with HDF5 on for me unfortunately. Should I upgrade the CI to test with HDF5?

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 3 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: :+1:

File: singularity-opac/neutrinos/non_cgs_neutrinos.hpp (hunk 5 above)
Comment (brryan) [reply singularity-opac/neutrinos/non_cgs_neutrinos.hpp:296]: I'd rather leave it this way so one can get CGS units back by default without having to think about a units modifier at all.

File: singularity-opac/neutrinos/spiner_opac_neutrinos.hpp (hunk 6 above)
Comment (brryan) [reply singularity-opac/neutrinos/spiner_opac_neutrinos.hpp:64]: I think the 10 is reasonable just to get a bit away from the "quantum foam" right at `::min()`, but I don't have great intuition here, if you think the 10 is bad I can remove it.

File: singularity-opac/photons/gray_opacity_photons.hpp (hunk 8 above)
Comment (brryan) [reply singularity-opac/photons/gray_opacity_photons.hpp:47]: I took off the `PerNu` because unlike e.g. `EmissivityPerNu` there's no `d \nu` in the denominator here. It's still a frequency-dependent call to the absorption coefficient though. Should I go back?

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 2 above)
Comment (Yurlungur) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: Worth testing it, I think.

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 4 above)
Comment (Yurlungur) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: But this doesn't matter for that, since the lambda parameter passed into the constructor is tied to the opacity object, not the MeanOpacity. I think this function, and the `nlambda_` can go.

File: singularity-opac/neutrinos/non_cgs_neutrinos.hpp (hunk 5 above)
Comment (Yurlungur) [reply singularity-opac/neutrinos/non_cgs_neutrinos.hpp:296]: Ok that's fine with me.

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 1 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp:32]: I remember why we have the `#define`... #17 I'm not sure if there's a good way around that (or a good way to have only one definition of that since you'd need an un-`undef`-ed `define` in a header file)

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 2 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: CI upgraded

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 4 above)
Comment (brryan) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp]: OK I see yes you would only use `nlambda()` here if you wanted to see how many parameters you called the opacity with earlier, you dont actually need it to size arrays or anything during runtime. OK I'll get rid of this.

File: singularity-opac/neutrinos/thermal_distributions_neutrinos.hpp (hunk 7 above)
Comment (brryan) [reply singularity-opac/neutrinos/thermal_distributions_neutrinos.hpp:30]: As above I don't see a clear path forward for shared `EPS` at the moment due to C++14

File: singularity-opac/neutrinos/mean_opacity_neutrinos.hpp (hunk 1 above)
Comment (Yurlungur) [reply singularity-opac/neutrinos/mean_opacity_neutrinos.hpp:32]: oooh. Ok. I guess we can move once we move to C++17.

**Review by Yurlungur — APPROVED** [review]
